### PR TITLE
HATCH-124: Missing field and not-null error codes

### PR DESCRIPTION
**Pull request by roy-coder**
*  Updating the `ValidationError` to use our new errors
* A little bit of refactoring around attribute types to make it more scalable
* Adding E2E test for HATCH-199 to fix a bug with it

### CODE REVIEW

**Review by abpin — COMMENTED** [review]
Looks good to me, better to have a second pair of eyes on it since I am still not familiar with the codebase.

**Review by bmomberger-bitovi — APPROVED** [review]
Approved, pending tests passing after main merge.

**Pull request merged**
